README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# fRegression

[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/paulnorthrop/fRegression?branch=main&svg=true)](https://ci.appveyor.com/project/paulnorthrop/fRegression)
[![R-CMD-check](https://github.com/paulnorthrop/fRegression/workflows/R-CMD-check/badge.svg)](https://github.com/paulnorthrop/fRegression/actions)
[![Coverage
Status](https://codecov.io/github/paulnorthrop/fRegression/coverage.svg?branch=main)](https://codecov.io/github/paulnorthrop/fRegression?branch=main)
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/fRegression)](https://cran.r-project.org/package=fRegression)
[![Downloads
(monthly)](https://cranlogs.r-pkg.org/badges/fRegression?color=brightgreen)](https://cran.r-project.org/package=fRegression)
[![Downloads
(total)](https://cranlogs.r-pkg.org/badges/grand-total/fRegression?color=brightgreen)](https://cran.r-project.org/package=fRegression)

## Rmetrics - Modelling Extreme Events in Finance

The **fRegression** package is a collection of functions for linear and
non-linear regression modelling. It implements a wrapper for several
regression models available in the base and contributed packages of R.

### An example

The following code simulates some regression data and fits various
models to these data.

``` r
library(fRegression)
# Simulate data: the response is linearly related to 3 explanatory variables 
x <- regSim(model = "LM3", n = 100)
  
# Linear modelling       
regFit(Y ~ X1 + X2 + X3, data = x, use = "lm") 
#> 
#> Title:
#>  Linear Regression Modeling 
#> 
#> Formula:
#>  Y ~ X1 + X2 + X3
#> 
#> Family:
#>  gaussian identity 
#> 
#> Model Parameters:
#> (Intercept)           X1           X2           X3  
#>     0.01578      0.73967      0.25128     -0.50611

# Robust linear modelling    
regFit(Y ~ X1 + X2 + X3, data = x, use = "rlm") 
#> 
#> Title:
#>  Robust Linear Regression Modeling 
#> 
#> Formula:
#>  Y ~ X1 + X2 + X3
#> 
#> Family:
#>  gaussian identity 
#> 
#> Model Parameters:
#> (Intercept)           X1           X2           X3  
#>     0.01968      0.74264      0.24736     -0.50123

# Generalised additive modelling       
regFit(Y ~ X1 + X2 + X3, data = x, use = "gam")  
#> 
#> Title:
#>  Generalized Additive Modeling 
#> 
#> Formula:
#>  Y ~ X1 + X2 + X3
#> 
#> Family:
#>  gaussian identity 
#> 
#> Model Parameters:
#> (Intercept)           X1           X2           X3  
#>     0.01578      0.73967      0.25128     -0.50611

# Projection pursuit modelling
regFit(Y ~ X1 + X2 + X3, data = x, use = "ppr") 
#> 
#> Title:
#>  Projection Pursuit Regression 
#> 
#> Formula:
#>  Y ~ X1 + X2 + X3
#> 
#> Family:
#>  gaussian identity 
#> 
#> Model Parameters:
#> -- Projection Direction Vectors --
#>        term 1     term 2
#> X1  0.7950116 -0.4422500
#> X2  0.2733278 -0.4863312
#> X3 -0.5415242 -0.7535894
#> -- Coefficients of Ridge Terms --
#>    term 1    term 2 
#> 0.9163087 0.0439332

# Feed-forward neural network modelling   
regFit(Y ~ X1 + X2 + X3, data = x, use = "nnet") 
#> 
#> Title:
#>  Feedforward Neural Network Modeling 
#> 
#> Formula:
#>  Y ~ X1 + X2 + X3
#> 
#> Family:
#>  gaussian identity 
#> 
#> Model Parameters:
#>    a 3-2-1 network with 11 weights
#>    options were - linear output units 
#>  [1]  3.3664690  0.5597762  0.2646774 -0.5300914  0.8276914 -0.4493467
#>  [7] -0.1400424  0.2787105 -0.5420174  5.4429808 -6.7838054

# Polychotonous Multivariate Adaptive Regression Splines
regFit(Y ~ X1 + X2 + X3, data = x, use = "polymars")
#>          1          2          3          4          5          6 
#>  0.9145273  1.1607611  1.0482997 -0.5673597 -0.4692621 -1.3336450 
#>           X1          X2          X3
#> 1  1.8197351 -0.39077723  0.24075985
#> 2  1.3704395  0.39665330 -0.02049151
#> 3  1.1963182  0.78156956  0.29685497
#> 4 -0.4068792 -0.01912605  0.55061347
#> 5 -0.6109788 -1.94431293 -0.71396821
#> 6 -1.5089120 -0.24550669  0.38003407
#> 
#> Title:
#>  Polytochomous MARS Modeling 
#> 
#> Formula:
#>  Y ~ X1 + X2 + X3
#> 
#> Family:
#>  gaussian identity 
#> 
#> Model Parameters:
#>   pred1 knot1 pred2 knot2       coefs          SE
#> 1     0    NA     0    NA  0.01577838 0.009803798
#> 2     1    NA     0    NA  0.73967249 0.009930477
#> 3     3    NA     0    NA -0.50611270 0.010729997
#> 4     2    NA     0    NA  0.25127670 0.010419817
```

### Installation

To get the current released version from CRAN:

``` r
install.packages("fRegression")
```


NEWS.md
# fRegression 4021.83: revision 6281 on R-Forge

* New maintainer: Paul Northrop
* Updated DESCRIPTION with links, moved timeDate, timeSeries, fBasics from Depends: to Imports: and moved MASS from Suggests to Imports.
* Removed Lazydata from DESCRIPTION: there are no data.
* Corrected some minor typos.

# fRegression 3042.82 and older versions

See file ChangeLog on [the CRAN fRegression page](https://CRAN.R-project.org/package=fRegression) for changes before 4021.83.


R/zzz.R

# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU Library General
# Public License along with this library; if not, write to the
# Free Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA  02111-1307  USA


###############################################################################


.onAttach <- 
  function(libname, pkgname)
{
  # do whatever needs to be done when the package is loaded
  # some people use it to bombard users with 
  # messages using 
  
  # packageStartupMessage( "\n" )
  # packageStartupMessage( "Rmetrics Package fRegression" ) 
  # packageStartupMessage( "Regression Based Decision and Prediction" )
  # packageStartupMessage( "Copyright (C) 2005-2014 Rmetrics Association Zurich" )  
  # packageStartupMessage( "Educational Software for Financial Engineering and Computational Science" ) 
  # packageStartupMessage( "Rmetrics is free software and comes with ABSOLUTELY NO WARRANTY." ) 
  # packageStartupMessage( "https://www.rmetrics.org --- Mail to: info@rmetrics.org" )
}


###############################################################################




inst/unitTests/runit.terms.R

# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU Library General
# Public License along with this library; if not, write to the
# Free Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA  02111-1307  USA


###############################################################################
# FUNCTION:             REGRESSION MODELLING DESCRIPTION:
#  regFit                Wrapper Function for Regression Models
###############################################################################


test.lmFit <- 
    function()
{   
    # Simulate Artificial LM:
    x <- regSim(model = "LM3", n = 50)
    
    # Fit Parameters:
    lmfit <- regFit(Y ~ X1 + X2 + X3, data = x, use = "lm") 
    lm <- stats::lm(Y ~ X1 + X2 + X3, data = x) 
    
    # Terms:
    terms(lmfit@fit)
    terms(lm)
 
    # Return Value:
    return()
}


# -----------------------------------------------------------------------------


test.rlmFit <- 
    function()
{   
    # Simulate Artificial LM:
    x = regSim(model = "LM3", n = 50)
    
    # Fit Parameters:
    rlmfit <- regFit(Y ~ X1 + X2 + X3, data = x, use = "rlm") 
    rlm <- MASS::rlm(Y ~ X1 + X2 + X3, data = x) 
    
    # Terms:
    terms(rlmfit@fit)
    terms(rlm)
      
    # Return Value:
    return()
}


# -----------------------------------------------------------------------------


test.glmFit <- 
    function()
{   
    # Simulate Artificial LM:
    x = regSim(model = "LM3", n = 50)
    
    # Fit Parameters:
    glmfit <- regFit(Y ~ X1 + X2 + X3, data = x, use = "glm") 
    glm <- stats::glm(Y ~ X1 + X2 + X3, data = x) 
    
    # Terms:
    terms(glmfit@fit)
    terms(glm)
     
    # Return Value:
    return()
}


# -----------------------------------------------------------------------------


test.gamFit <- 
    function()
{   
    # Simulate Artificial LM:
    x = regSim(model = "LM3", n = 50)
    
    # Fit Parameters:
    gamfit <- regFit(Y ~ s(X1) + s(X2) + X3, data = x, use = "gam") 
    gam <- mgcv::gam(Y ~ X1 + X2 + X3, data = x) 
    
    # Terms:
    terms(gamfit@fit)
    terms(gam)
      
    # Return Value:
    return()
}


# -----------------------------------------------------------------------------


test.pprFit <- 
    function()
{   
    # Simulate Artificial LM:
    x <- regSim(model = "LM3", n = 50)
    
    # Fit Parameters:
    pprfit <- regFit(Y ~ X1 + X2 + X3, data = x, use = "ppr") 
    ppr <- stats::ppr(Y ~ X1 + X2 + X3, data = x, nterms = 2) 
    
    # Terms:
    terms(pprfit@fit)
    terms(ppr)
       
    # Return Value:
    return()
}


# -----------------------------------------------------------------------------


test.nnetFit <- 
    function()
{   
    # Simulate Artificial LM:
    x <- regSim(model = "LM3", n = 50)
    
    # Fit Parameters:
    nnetfit <- regFit(Y ~ X1 + X2 + X3, data = x, use = "nnet") 
    nnet <- nnet::nnet(Y ~ X1 + X2 + X3, data = x, trace = FALSE, 
        size = 2, linout = TRUE) 
    
    # Terms:
    terms(nnetfit@fit)
    terms(nnet)
       
    # Return Value:
    return()
}


# -----------------------------------------------------------------------------


test.polymarsFit <- 
    function()
{   
    # Simulate Artificial LM:
    x = regSim(model = "LM3", n = 50)
    
    # Fit Parameters:
    polymarsfit <- regFit(Y ~ X1 + X2 + X3, data = x, use = "polymars") 
    polymars <- fRegression:::.polymars(Y ~ X1 + X2 + X3, data = x)
    
    # Terms:
    terms(polymarsfit@fit)
    terms(polymars)
    
    # Return Value:
    return()
}


###############################################################################




inst/unitTests/runit.polymars.R

# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU Library General
# Public License along with this library; if not, write to the
# Free Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA  02111-1307  USA


################################################################################
# FUNCTION:             REGRESSION MODELLING DESCRIPTION:
#  polymars              Polymars Regression
#  .polymars             Polymars regress from package polspline
#  .polymarsDefault      Internal Function
#  .polymarsVormula      Internal Function
#  .predict.polymars     Internal Function
################################################################################


test.polymars <- 
    function()
{   
    # Simulate Artificial LM:
    x <- regSim(model = "LM3", n = 50)
    
    # Original Polymars:
    fit <- polspline::polymars(responses = x[,1], predictors = x[,2:4])
    
    # Model Fitting:
    fit$fitting
    
    # Model Produced:
    fit$model
    fit$coef
    
    # Summary:
    #   Note print.polymars = summary.polymars
    polspline::summary.polymars(fit)
    
    # Predict:
    ans <- polspline::predict.polymars(object = fit, x = x[,-1]) 
    as.vector(ans)
    as.vector(fit$fitted)
    
    # Return Value:
    return()
}


# ------------------------------------------------------------------------------


test.polymarsDefault <- 
    function()
{   
    # Simulate Artificial LM:
    set.seed(4711)
    x = regSim(model = "LM3", n = 50)
    
    # Polymars Wrapper:
    fit1 = fRegression:::.polymarsDefault(responses = x[,1], predictors = x[, 2:4])
    class(fit1)
    names(fit1)
    
    # Note, this fails:
    # fit1 = .polymars(responses = x[,1], predictors = x[,2:4])
    
    # Model Fitting:
    fit1$fitting
    
    # Model Produced:
    #   fit1$model reserved for model series, use ...
    fit1$coef
    
    # Summary:
    print(fit1)
    
    # Print:
    summary(fit1)
    
    # Predict:
    ans <- polspline::predict.polymars(object = fit1, x = x[,-1]) 
    as.vector(ans)
    as.vector(fit1$fitted)
    
    # Check:
    fit1$ranges.and.medians 
       
    # Return Value:
    return()
}


# -----------------------------------------------------------------------------


test.polymarsFormula <- 
    function()
{   
    # Simulate Artificial LM:
    set.seed(4711)
    x <- regSim(model = "LM3", n = 50)
    
    # Polymars Formula Wrapper:
    fit2 <- fRegression:::.polymarsFormula(formula = Y ~ X1 + X2 + X3, data = x)
    fit2 <- fRegression:::.polymars(formula = Y ~ X1 + X2 + X3, data = x)
    class(fit2)
    names(fit2)
    
    # Model Fitting:
    fit2$fitting
    
    # Model Produced:
    #   fit$model reserved for model series, use ...
    fit2$coef
    
    # Summary:
    print(fit2)
    
    # Print:
    summary(fit2)
    
    # Predict:
    fit2$model <- fit2$coef
    ans <- polspline::predict.polymars(object = fit2, x = x[,-1]) 
    as.vector(ans)
    as.vector(fit2$fitted)
    
    # Check:
    fit2$ranges.and.medians 
    
    # Return Value:
    return()
}

# -----------------------------------------------------------------------------


test.regFit.polymars <- 
    function()
{   
    # Simulate Artificial LM:
    x <- regSim(model = "LM3", n = 50)
    
    # Polymars Formula Wrapper:
    fit <- regFit(formula = Y ~ X1 + X2 + X3, data = x, use = "polymars")
    class(fit)
    
    # Model Fitting:
    fit@fit$fitting
    
    # Model Produced:
    #   fit$model reserved for model series, use ...
    fit@fit$coef
    
    # Summary:
    print(fit)
    
    # Print:
    summary(fit)
           
    # Return Value:
    return()
}


# -----------------------------------------------------------------------------


test.regFit.polymars.methods <- 
    function()
{   
    # Simulate Artificial LM:
    x <- regSim(model = "LM3", n = 20)
    
    # Polymars Formula Wrapper:
    polymarsfit <- regFit(
      formula = Y ~ X1 + X2 + X3, data = x, use = "polymars")
   
    # Print:
    print(polymarsfit)
    
    # Summary:
    summary(polymarsfit)
    
    # Fitted Values:
    fitted(polymarsfit)
    slot(polymarsfit, "fitted")
    
    # Residuals:
    residuals(polymarsfit)
    slot(polymarsfit, "residuals")
    
    # Coefficients:
    coef(polymarsfit)
    
    # Formula
    formula(polymarsfit)
    
    # Return Value:
    return()
}


# -----------------------------------------------------------------------------


test.regFit.polymars.predict <- 
    function()
{   
    # Simulate Artificial LM:
    x <- regSim(model = "LM3", n = 50)
    
    # regFit / Polymars Formula Wrapper:
    fit <- regFit(formula = Y ~ X1 + X2 + X3, data = x, use = "polymars")
    class(fit)
    fit@fit$cmd
     
    # Predict from predict.polymars:
    object <- fit@fit
    class(object) = "polymars"
    object
    object$model = object$coef
    ans <- polspline::predict.polymars(object = object, x = x[,-1]) 
    as.vector(ans)
    as.vector(fit@fitted)
    
    # Predict from predict.fREG:
    ans <- predict(object = fit, newdata = x) 
    as.vector(ans)
    as.vector(fit@fitted)
    
    # Return Value:
    return()
}


###############################################################################




inst/unitTests/runit.TermPlots.R

# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU Library General
# Public License along with this library; if not, write to the
# Free Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA  02111-1307  USA


################################################################################
# FUNCTION:                 REGRESSION TERM PLOTS
#  termPlot                  Line Plot
#  termPersp                 Perspective Plot
#  termContour               Contour Plot
################################################################################


test.termPlot <-
    function()
{

    x <- regSim(model = "LM3", n = 100)
    lmfit <- regFit(Y ~ X1 + X2 + X3, data = x, use = "lm")


    # Simulate Data - a data frame:
    DATA <- regSim(model = "GAM3", n = 100)
    head(DATA)
    class(DATA)

    # Convert to a timeSeries object:
    DATATS <- as.timeSeries(DATA)
    head(DATATS)
    class(DATATS)
    
    require(mgcv)
    
    # Fit:
    LM    = regFit(Y ~ 1 + X1 + X2 + X3, data = DATATS, use = "lm")
    RLM   = regFit(Y ~ 1 + X1 + X2 + X3, data = DATATS, use = "rlm")
    AM    = regFit(Y ~ 1 + X1 + X2 + X3,   DATATS, use = "gam")
    PPR   = regFit(Y ~ X1 + X2 + X3,     data = DATATS, use = "ppr")
    PPR4  = regFit(Y ~ X1 + X2 + X3,     data = DATATS, use = "ppr", nterms = 4)
    POLYMARS = regFit(Y ~ X1 + X2 + X3,     data = DATATS, use = "polymars")
    NNET  = regFit(Y ~ X1 + X2 + X3,     data = DATATS, use = "nnet")
    NNET6 = regFit(Y ~ X1 + X2 + X3,     data = DATATS, use = "nnet", size = 6)
    
    ## TODO: Term Plot:
    ## par(ask = FALSE)
    ## par(mfrow = c(1, 1))
    ## termPlot(LM)
    ## termPlot(RLM)
    ## termPlot(AM)
    ## termPlot(PPR)
    ## termPlot(POLYMARS)
    ## termPlot(NNET)

    ## TODO:
    ## par(ask = FALSE)
    ## par(mfrow = c(1, 1))
    ## termPlot(LM, terms = "X1")
    ## termPlot(RLM, terms = "X1")
    ## termPlot(AM, terms = "X1")
    ## termPlot(PPR, terms = "X1")
    ## termPlot(PPR4, terms = "X1")
    ## termPlot(POLYMARS, terms = "X1")
    ## termPlot(NNET, terms = "X1")
    ## termPlot(NNET6, terms = "X1")

    # Return Value:
    return()
}


# ------------------------------------------------------------------------------


test.termPersp <-
    function()
{
    # Simulate Data - a data frame:
    DATA <- regSim(model = "GAM3", n = 100)
    head(DATA)
    class(DATA)
    # Convert to a timeSeries object:
    DATATS <- as.timeSeries(DATA)
    head(DATATS)
    class(DATATS)
    
    require(mgcv)
    
    # Fit:
    LM    = regFit(Y ~ 1 + X1 + X2 + X3, data = DATATS, use = "lm")
    RLM   = regFit(Y ~ 1 + X1 + X2 + X3, data = DATATS, use = "rlm")
    AM    = regFit(Y ~ 1 + s(X1)+s(X2)+s(X3),   DATATS, use = "gam")
    PPR   = regFit(Y ~ X1 + X2 + X3,     data = DATATS, use = "ppr")
    PPR4  = regFit(Y ~ X1 + X2 + X3,     data = DATATS, use = "ppr", nterms = 4)
    POLYMARS = regFit(Y ~ X1 + X2 + X3,  data = DATATS, use = "polymars")
    NNET  = regFit(Y ~ X1 + X2 + X3,     data = DATATS, use = "nnet")
    NNET6 = regFit(Y ~ X1 + X2 + X3,     data = DATATS, use = "nnet", size = 6)
    
    ## TODO:  Bivariate Perspective Term Plot:
    ## par(ask = FALSE)
    ## par(mfrow = c(1, 1))
    ## termPersp(LM,    terms = c("X1", "X2"))
    ## termPersp(RLM,   terms = c("X1", "X2"))
    ## termPersp(AM,    terms = c("X1", "X2"))
    ## termPersp(PPR,   terms = c("X1", "X2"))
    ## termPersp(PPR4,  terms = c("X1", "X2"))
    ## termPersp(POLYMARS, terms = c("X1", "X2"))
    ## termPersp(NNET,  terms = c("X1", "X2"))
    ## termPersp(NNET6, terms = c("X1", "X2"))
    
    # Return Value:
    return()
}


# ------------------------------------------------------------------------------


test.termContour <-
    function()
{

    # Simulate Data - a data frame:
    DATA = regSim(model = "GAM3", n = 100)
    head(DATA)
    class(DATA)
    # Convert to a timeSeries object:
    DATATS = as.timeSeries(DATA)
    head(DATATS)
    class(DATATS)
    
    require(mgcv)
    
    # Fit:
    LM    = regFit(Y ~ 1 + X1 + X2 + X3, data = DATATS, use = "lm")
    RLM   = regFit(Y ~ 1 + X1 + X2 + X3, data = DATATS, use = "rlm")
    AM    = regFit(Y ~ 1 + s(X1)+s(X2)+s(X3),   DATATS, use = "gam")
    PPR   = regFit(Y ~ X1 + X2 + X3,     data = DATATS, use = "ppr")
    PPR4  = regFit(Y ~ X1 + X2 + X3,     data = DATATS, use = "ppr", nterms = 4)
    POLYMARS = regFit(Y ~ X1 + X2 + X3,     data = DATATS, use = "polymars")
    NNET  = regFit(Y ~ X1 + X2 + X3,     data = DATATS, use = "nnet")
    NNET6 = regFit(Y ~ X1 + X2 + X3,     data = DATATS, use = "nnet", size = 6)
    
    ## TODO: Bivariate Contour Term Plot:
    ## par(ask = FALSE)
    ## par(mfrow = c(1, 1))
    ## termContour(LM,    terms = c("X1", "X2"))
    ## termContour(RLM,   terms = c("X1", "X2"))
    ## termContour(AM,    terms = c("X1", "X2"))
    ## termContour(PPR,   terms = c("X1", "X2"))
    ## termContour(PPR4,  terms = c("X1", "X2"))
    ## termContour(POLYMARS, terms = c("X1", "X2"))
    ## termContour(NNET,  terms = c("X1", "X2"))
    ## termContour(NNET6, terms = c("X1", "X2"))
    
    # Return Value:
    return()
}


# ------------------------------------------------------------------------------


test.termComparison <- 
    function()
{
    # Simulate Data - a data frame:
    DATA = regSim(model = "GAM3", n = 100)
    head(DATA)
    class(DATA)
    
    # Convert to a timeSeries object:
    DATATS = as.timeSeries(DATA)
    head(DATATS)
    class(DATATS)
    
    require(mgcv)
    
    ## TODO:
    if (FALSE) {
    
        # Comparison:
        par(ask = FALSE)
        par(mfrow = c(1, 1))
    
        LM = regFit(Y ~ 1 + X1 + X2 + X3, data = DATATS, use = "lm")
        termPlot(LM)
    
        AM = regFit(Y ~ 1 + s(X1)+s(X2)+s(X3), data = DATATS, use = "gam")
        termPlot(AM)
        am = gam(formula = Y ~ s(X1) + s(X2) + s(X3), data = DATA)
        for (s in 1:3) {
            plot(am, residuals = residuals(am), se = TRUE,
                main = "AM", cex = 0.7, select = s, pch = 19); grid()
        }
    }

    # Return Value:
    return()
}


###############################################################################




R/regFit.R

# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU Library General
# Public License along with this library; if not, write to the
# Free Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA  02111-1307  USA


###############################################################################
# FUNCTION:             PARAMETER ESTIMATION:
#  regFit                Wrapper Function for Regression Models
#  .lmFit                 Linear Regression Model
#  .rlmFit                Robust Linear Regression Model
#  .glmFit                Generalized Linear Model
#  .gamFit                Generalized Additive Model
#  .pprFit                Projection Pursuit Regression Model
#  .nnetFit               Feedforward Neural Network Model
#  .polymarsFit           Polychotomous MARS Model
###############################################################################


###############################################################################
# MODEL:        PACKAGE     print   plot   summary   print     predict
#                                    persp           summary
#   lm          stats       x       x      x         x         x
#   rlm         MASS
#   glm         stats       x       -      x         x         x
#   gam         mgcv        x       x      x         x         x
#   ppr         stats       x       x      x         x         x
#   nnet        nnet        x       -      x         x         x
#   polymars*   polspline   -       xx     x         -         x
###############################################################################


regFit <-
    function (formula, data, family = gaussian,
    use = c("lm", "rlm", "glm", "gam", "ppr", "nnet", "polymars"),
    title = NULL, description = NULL, ...)
{
    # A function implemented by Diethelm Wuertz

    # Description:
    #   Common function call for several selected regression models.

    # Details:
    #   This is a wrapper function for the following regrssion models:
    #     LM          Linear Regression Modelling
    #     RLM         Robust Linear Regression Modelling
    #     GLM         Generalized Linear Modelling
    #     GAM         Generalized Additive Modelling
    #     PPR         Projection Pursuit Regression
    #     NNET        Feedforward Neural Net
    #     POLYMARS    Polychotomous MARS Modeling

    # Notes:
    #   Available Methods are
    #   "print", "plot", "summary", "predict"
    #   "coef", "formula", "residuals" "fitted", "vcov"

    # Example:
    #   regFit(Y ~ X1 + X2 + X3, regSim())

    # FUNCTION:

    # Match Arguments:
    use <- match.arg(use)
    if (missing(data)) data <- NULL

    # Transform data into a dataframe
    if (!is.null(data)) {
      Data <- if (inherits(data, "timeSeries")) data else as.timeSeries(data)
      data <- as.data.frame(data)
    } else {
      Data <- data <- NULL
    }

    # Function to be called:
    fun <- paste(".", match.arg(use), sep = "")

    # Title:
    if (is.null(title)) {
        if (use == "lm") title = "Linear Regression Modeling"
        if (use == "rlm") title = "Robust Linear Regression Modeling"
        if (use == "glm") title = "Generalized Linear Modeling"
        if (use == "gam") title = "Generalized Additive Modeling"
        if (use == "ppr") title = "Projection Pursuit Regression"
        if (use == "nnet") title = "Feedforward Neural Network Modeling"
        if (use == "polymars") title = "Polytochomous MARS Modeling"
    }

    # Description:
    if (is.null(description)) description = description()

    # Compose Command to be Called:
    cmd <- match.call()
    if (!is.null(cmd$use)) cmd = cmd[-match("use", names(cmd), 0)]
    cmd[[1]] <- as.name(fun)
    # Use this to access hidden functions in a parent frame:
    #cmd[[1]] <- substitute(fRegression:::f, list(f=as.name(fun)))

    # Ensure that data is a data.frame
    if (!is.null(cmd$data)) cmd$data <- as.name("data")
    # Use this to directly pass the argument from the parent frame:
    #if (!is.null(cmd$data)) cmd$data <- call("as.data.frame", cmd$data)

    # Fit Regression Model:
    fit <- eval(cmd)
    # Use this to evaluate in parent frame:
    #fit <- eval(cmd, parent.frame())

    # Add "cmd" to Fit:
    fit$cmd <- cmd

    # Add "xlevels" to Fit (if missing):
    if (is.null(fit$xlevels)) fit$xlevels = list()

    # Add "residuals" and "fitted" to Fit (to be sure ...):
    fit$residuals <- as.vector(fit$residuals)
    fit$fitted.values <- as.vector(fit$fitted.values)

    # Add "parameters" as Alternative:
    fit$parameters <- fit$coef

    # Extend to class "list":
    class(fit) <- c("list", class(fit))
    if (!inherits(fit, "lm")) class(fit) = c(class(fit), "lm")

    # Return Value:
    new("fREG",
        call = as.call(match.call()),
        formula = as.formula(formula),
        family = as.character(gaussian()),
        method = use,
        # data is as.data.frame(data), Data is as.timeSeries(data):
        data = list(data = data, Data = Data),
        fit = fit,
        residuals = fit$residuals,
        fitted = fit$fitted.values,
        title = as.character(title),
        description = as.character(description)
    )
}


###############################################################################


.lm <- 
  function(...)
{
  stats::lm(...)
}


# -----------------------------------------------------------------------------


.rlm <- 
  function(...)
{
  MASS::rlm(...)
}


# -----------------------------------------------------------------------------


.glm <- 
  function(...)
{
  stats::glm(...)
}


# -----------------------------------------------------------------------------


.gam <- 
  function(...)
{
  mgcv::gam(...)
}


# -----------------------------------------------------------------------------


.ppr <- 
  function(..., nterms = 2)
{
  stats::ppr(..., nterms = nterms)
}


# -----------------------------------------------------------------------------


.nnet <- 
  function(..., trace = FALSE, size = 2, linout = TRUE)
{
  nnet::nnet(..., trace = trace, size = size, linout = linout)
}


# -----------------------------------------------------------------------------


.polymars <- 
  function(...) 
{
  .polymarsFormula(...)
}


###############################################################################


